tiatoolbox/dataloader/wsireader.py

```diff
@@ -82,39 +91,524 @@ def slide_info(self):
+    ):
+        """Find optimal parameters for reading bounds at a given resolution.
+
         Args:
             start_w (int): starting point in x-direction (along width)
+                in baseline (level 0) reference frame.
             start_h (int): starting point in y-direction (along height)
+                in baseline (level 0) reference frame.
             end_w (int): end point in x-direction (along width)
+                in baseline (level 0) reference frame.
             end_h (int): end point in y-direction (along height)
-            level (int): pyramid level to read the image
+                in baseline (level 0) reference frame.
+            resolutions (float): desired output resolution
+            units (str): the units of scale, default = "level".
+                Supported units are: microns per pixel (mpp), objective
+                power (power), pyramid / resolution level (level),
+                pixels per baseline pixel (baseline).
+            precision (int, optional): Decimal places to use when
+                finding optimal scale. See
+                :func:`find_optimal_level_and_downsample` for more.
 
         Returns:
-            img_array : ndarray of size MxNx3
-            M=end_h-start_h, N=end_w-start_w
+            (int, tuple of int, tuple of int, float): Read parameters of
+                optimal read level, bounds (start_w, start_h, end_w,
+                end_h) of the region in the optimal level reference
+                frame, correct size to output after reading and applying
+                downscaling, downscaling factor to apply after reading
+                to get the correct output size and resolution.
+        """
+        read_level, post_read_scale_factor = self._find_optimal_level_and_downsample(
+            resolution, units, precision
+        )
+        info = self.slide_info
+        level_downsample = info.level_downsamples[read_level]
+        location = np.array([start_w, start_h])
+        size = np.array([end_w - start_w, end_h - start_h])
+        level_size = np.round(np.array(size) / level_downsample).astype(int)
+        level_location = np.round(location / level_downsample).astype(int)
+        level_bounds = (*level_location, *(level_location + level_size))
+        output_size = np.round(level_size * post_read_scale_factor).astype(int)
+        return read_level, level_bounds, output_size, post_read_scale_factor
+
+    def read_rect(self, location, size, resolution=0, units="level"):
+        """Read a region of the whole slide image at a location and size.
+
+        Location is in terms of the baseline image (level 0  /
+        maximum resolution), and size is the output image size.
+
+        Reads can be performed at different resolutions by supplying a
+        pair of arguments for the resolution and the
+        units of resolution.
+
+        The field of view varies with resolution. For a fixed
+        field of view see :func:`read_bounds`.
```

Comment:
Is the hyperlink fixed?

Reply:
Yep

Reply:
See commit https://github.com/TIA-Lab/tiatoolbox/pull/40/commits/57af1864939bfaa942e47c7387fb791fe7aa5b29